Afterlife Sketch Boilerplate Code

prodestrian opened the pull request:
Add initial Arduino sketch with basic library implementation

* Add `ConfigManager` to read, deserialize, and process JSON Config from LittleFS Filesystem
* Add `Lights` class to handle interaction with LEDs via `FastLED`
* Add `Settings` class to define configuration structures
* Add sample `config.json`
* Write initial `AfterlifeLightKit.ino` sketch
* Bump `.editorconfig` `indent_size` from `2` to `4` for readability


## Testing
To test this out, assuming your pin assignments haven't changed (Powercell on `GPIO0` and Cyclotron on `GPIO5`):
1. Upload the `config.json` to the LittleFS filesystem
2. Upload the sketch to the device

It should immediately start looping through the Afterlife animations on the Powercell and Cyclotron.

Try modifying some of the Afterlife variables in `config.json`, upload to the device, wait for it to auto-reboot, and the changes should appear. For example, change the RGB colours, the speeds (intervals), or the Cyclotron direction (forward/reverse).

## Notes/Caveats
1. This doesn't include Wifi but I think we've figured out the basics there. I'm reluctant to bring that feature into the codebase until we've cleaned up what we have for now.
2. There are a lot of TODO notes highlighting many areas where I'm still workout out what we need.
3. Pin assignments are hard-coded. I couldn't change them via Configuration, [FastLED](https://github.com/FastLED/FastLED/issues/1137) doesn't like this. If someone is using custom hardware or changing the pin assignments, they may need to edit the `Settings.h` and recompile. I know WLED can change pin assignments via web interface, but it's a huge amount of troubleshooting for a feature that probably won't get used.
4. The `ConfigManager` needs some rethinking. It works fine but it's already got way out of hand. I don't know enough about C code to know how we clean this up. I'd really like to grab the JSON document, push it into a data converter/processor, and have it return a Configuration object. I don't know how to do that in C, and I gave up trying to find an example which made sense. Maybe we just have to handle every single configuration option by hand, but that isn't ideal if we're going to end up with dozens.
5. The code only supports Afterlife-mode at the moment, when we abstract the animations to separate classes we can implement Classic mode.

## Interaction
I haven't had a chance to connect inputs to this yet, but I'm assuming all the logic will take place inside the main sketch. We may want to abstract the inputs to their own class, which can handle reading them into a current State (not sure if they need to do debouncing, we can use `Bounce2` if we need that). This can handle changing modes, firing, startup, shutdown, whatever is needed.


Anyway, I'm out of time for today so this is where I'll have to leave it.

prodestrian committed:
Afterlife Sketch Boilerplate Code

Add initial Arduino sketch with basic library implementation

* Add ConfigManager to read, deserialize, and process JSON Config
  from LittleFS Filesystem
* Add Lights class to handle interaction with LEDs via FastLED
* Add Settings class to define configuration structures
* Add sample config.json
* Write initial AfterlifeLightKit.ino sketch
* Bump editorconfig indent_size from 2 to 4 for readability

prodestrian commented on SOFTWARE/AfterlifeLightKit/ConfigManager.cpp:
```diff
@@ -0,0 +1,115 @@
+    //_configuration.classicSettings.powercellSpeed = doc["modes"]["classic"]["powercell"]["speed"] | 1000;
+
+
+    _configuration.afterlifeSettings.powercell.speed = doc["modes"]["afterlife"]["powercell"]["speed"] | 50;
```
We might be able to rewrite this to set all the `afterlifeSettings` at once from the JSON, especially if every "mode" has the exact same settings available (speed, color, direction, etc).

prodestrian commented on SOFTWARE/AfterlifeLightKit/Settings.h:
```diff
@@ -0,0 +1,65 @@
+    LightStrip Cyclotron = {60};
+};
+
+struct RGBColor {
```
Hoping we can eventually remove this in favour of CRGB from FastLED, but I couldn't get it working:
```ino
struct PowercellSettings
{
    Directions direction;
    int speed;
    CRGB color;
};
```
(I can't remember the exact compiler error I got)

prodestrian commented on SOFTWARE/AfterlifeLightKit/Lights.h:
```diff
@@ -0,0 +1,101 @@
+         * So 20 should be more than enough for the Powercell, and 100 for the Cyclotron
+         */
+        CRGB _powercellLEDS[20];
+        CRGB _cyclotronLEDS[100];
```
I didn't want to hard-code these LED strip lengths, but it seems to be required for memory allocation and can't be initialized dynamically from configuration, so I set them higher than what most people would need. I'm now thinking we just go back to `16` and `60` respectively, define them in `Settings.h`, and let people recompile if they need to. You can always change the software length from configuration, just not the hardware length.

ajquick replied:
I certainly think it is fine to choose a safe margin value for each. Should anyone need something way more, they should edit the software because they are gonna have something weird going on anyways.

ajquick committed:
Update ConfigManager.cpp

Static document size not large enough to handle current configuration. "Failed to read file."

Change to 1024 for current document size, but be aware of capacity for static JSON documents. 

http://arduinojson.org/v6/assistant

Added debug to output Json file to Serial.

ajquick reviewed (commented):
I found the 512 size for the Json document was not large enough and it wasn't actually opening the file and importing the settings. 1024 worked according to the calculator. http://arduinojson.org/v6/assistant

ajquick commented:
I should have asked before debugging. Does this code _work_?

I couldn't get anything output to the lights, but it certainly should have worked from everything I saw.

I added ```FastLED.setBrightness(255);``` just in case it was a missing brightness value but that didn't seem to work.

ajquick commented:
WHOOPS. I am dumb.

In order to test with my hardware, the logic converter must always be enabled first:

````
pinMode(15, OUTPUT);
digitalWrite(15, HIGH);
````

Added that and it works.

I am however experiencing an interesting phenomenon.

![Interesting Phenomenon](https://user-images.githubusercontent.com/6588235/177019394-ad764bff-6d47-462d-beed-78651bf6f419.gif)

The last 4 LEDs on each string occasionally show the other string's data. The last 4 LEDs on the Powercell are sometimes red and the last 4 LEDs on the Cyclotron are sometimes blue. Perhaps memory collisions or is this the flicker that was previously mentioned?

**Nevermind! This was because I had set the configuration to a number of LEDs larger than what would be allowed by the hardcoded assignment. I had 24 LEDs for my Powercell proxy, but it was hard capped at 20, so the last 4 were very erratic.**

ajquick committed:
Update Lights.cpp

This change forces the power cell to change direction once it reaches the end of the strip.

ajquick commented:
Added a simple bit of code to reverse the direction of the Powercell when it reaches the ends.

prodestrian commented:
Ah I hadn't imagined that we'd ever want more than 20 pixels for a powercell, but that makes sense.
I really tried to make this size dynamic but I couldn't get it working. If it was a local variable it wouldn't be an issue, but it has to be global which means I need to declare it (with the size) upfront. Someone with better C skills can probably tell me how we could work around this.

Maybe we should just move all hardware configuration out of the JSON and into `Settings.h`? At the very least you shouldn't need to edit `Lights.h` for simple hardware changes, it should be fairly generic.

My logic level converters were due to arrive Friday but they were delayed until Monday. Once I have them I'll add them in and connect to Pin 15 so I can finally mirror the setup you have :+1:

Good work on the Powercell animation change, I think my next task on top of this is to break out the modes/animations into separate classes so we can start adding features. I'm guessing we'll have one class for Powercell Animations, and one for Cyclotron Animations. And then we use Configuration to assign them to different States.
So if you _really_ want your Afterlife cyclotron to switch to a Reverse-1984 animation while overheating, you can configure that via JSON.

When I've done a few of these I can start work on the Classic/1984 Cyclotron mode, which will give me a chance to setup Offsets (so you don't need to worry about what angle your LED `0` is positioned at). I think I can code this manually but I saw some references to offsets in the FastLED code, would be great to do this natively and not even have to worry about it.
So, maybe your first LED should be `22`. Which means it should loop to LED `59` and then circle back to `0`. Less important for the Afterlife cyclotron but pretty important for Classic mode.

prodestrian commented:
Going to merge this now and pick up on the animations/events in a new branch.

prodestrian commented:
Pulled code into local environment, recompiled and uploaded to device with zero additional changes. Confirmed still working, but now with ascend/descend animation on Powercell :+1:

ajquick commented:
> When I've done a few of these I can start work on the Classic/1984 Cyclotron mode, which will give me a chance to setup Offsets (so you don't need to worry about what angle your LED `0` is positioned at). I think I can code this manually but I saw some references to offsets in the FastLED code, would be great to do this natively and not even have to worry about it. So, maybe your first LED should be `22`. Which means it should loop to LED `59` and then circle back to `0`. Less important for the Afterlife cyclotron but pretty important for Classic mode.

For the classic mode, it might be best to just require the explicit defining of which LEDs are to be used. I know there was another Arduino code library somewhere that had to address this and come up with a solution for it, but I'm not sure which one it was. Basically the location of the Cyclotron holes is not even and it is offset from the center of the Cyclotron. So the LED ring ends up at a weird position and then you have unequal distances between the holes.

Here is a diagram example with a 40 LED ring.

![Cyclotron](https://user-images.githubusercontent.com/6588235/177027650-a5bcfb85-4988-4b9c-91eb-97938849c472.png)

I suppose it could absolutely be possible to just turn on LEDs 4-9 for one hole, 14-19 for the next.. and so on. That could actually be done with scaling based on the number of LEDs on a given ring and the assumption that the start of the LED ring would always be at the top.

If you added in the offset, that could help by allowing the ring to be installed in any orientation.

I suspect people will want just 1 LED on at a time for 1984 mode though. So they would want to get very precise about what LED to turn on. Perhaps allow the specification of which LED to turn on for each of the 4 points and then add a width attribute so someone could choose to have just 1 LED or 3, 5, 7.. etc LED for each segment of the 1984 Cyclotron animation.

Keep in mind a 40 LED ring might be used for a full sized pack, but a 32 LED ring might be used for an 85% scaled pack. Explicitly assigning which LEDs to use in the config would be easier than perhaps calculating the ratio or angle between each Cyclotron light.

prodestrian commented:
I was just going to be even simpler about it, on my 60-LED ring I was going to light up 0-14, 15-29, 30-44, 45-60 (or some other starting point based on the offset calculation, eg 22-36, 37-51, etc).

With a good diffuser it's not too much of a problem, but if someone only wanted 1 or 2 LEDs illuminated at once it would be good to allow that, and as you said the offset wouldn't work with that because it could be off-center in each hole.

I'll have to think about exactly how we'll structure the configuration, I assume just an array of LED numbers. It could be handy to add a Preview/Test feature for this, so you can enter a list of numbers and then press Test in the web interface to see those illuminated. You can fine tune it in real time and get the LEDs lined up perfectly.

I'll come back to this (maybe as part of #3 ) once I've written some of the abstract classes.
